### PR TITLE
For #40884: Extend support for direct uploads to S3

### PR DESCRIPTION
**Pull request by daigles**
- Added multi-part uploads to support uploads of files > 5 GB

### CODE REVIEW

File: shotgun_api3/shotgun.py [hunk 1]
```diff
@@ -2059,10 +2059,6 @@ def share_thumbnail(self, entities, thumbnail_path=None, source_entity=None,
             "filmstrip_thumbnail" : filmstrip_thumbnail,
         }
 
-        params.update(self._auth_params())
-
-        # Create opener with extended form post support
-        opener = self._build_opener(FormPostHandler)
```
Comment (daigles) [inline shotgun_api3/shotgun.py]: removing dead code

File: shotgun_api3/shotgun.py [hunk 2]
```diff
@@ -2374,11 +2355,17 @@ def _get_attachment_upload_info(self, is_thumbnail, filename):
         if not str(upload_info).startswith("1"):
             raise ShotgunError("Could not get upload_url but " \
                                "not sure why.\nPath: %s\nUrl: %s\nError: %s" % (
-                                   path, url, str(upload_info)))
+                                   filename, url, str(upload_info)))
```
Comment (daigles) [inline shotgun_api3/shotgun.py]: bug fix: 'path' is actually not defined in this scope.

File: shotgun_api3/shotgun.py [hunk 3]
```diff
@@ -2364,7 +2344,8 @@ def _get_attachment_upload_info(self, is_thumbnail, filename):
             "filename" : filename
         }
 
-        params.update(self._auth_params())
```
Comment (daigles) [inline shotgun_api3/shotgun.py]: Moved this to the 'send_form' function.

File: shotgun_api3/shotgun.py [hunk 4]
```diff
@@ -2219,33 +2215,16 @@ def _upload_to_storage(self, entity_type, entity_id, path, field_name, display_n
 
         # Step 1: get the upload url
 
-        upload_info = self._get_attachment_upload_info(is_thumbnail, filename)
+        multipart_upload = (os.path.getsize(path)  > 20000000)
```
Comment (manneohrstrom) [inline shotgun_api3/shotgun.py]: nit: on the toolkit side we would probably prefix with is to indicate boolean, e.g. `is_multipart_upload`

File: shotgun_api3/shotgun.py [hunk 5]
```diff
@@ -2219,33 +2215,16 @@ def _upload_to_storage(self, entity_type, entity_id, path, field_name, display_n
...
-                raise ShotgunError("Server encountered an internal error.\n%s\n%s\n\n" % (url, e))
-            else:
-                raise ShotgunError("Unanticipated error occurred uploading %s: %s" % (path, e))
-        finally:
-            fd.close()
+        # Step 2: upload the file
 
-        LOG.debug("File uploaded to Cloud storage: %s", filename)
+        if multipart_upload:
+            self._multipart_upload_file_to_storage(path, upload_info)
```
Comment (manneohrstrom) [inline shotgun_api3/shotgun.py]: minor: some comments explaining why we need to make the distinction here based on file size could helpful.

File: shotgun_api3/shotgun.py [hunk 6]
```diff
@@ -2364,7 +2344,8 @@ def _get_attachment_upload_info(self, is_thumbnail, filename):
             "filename" : filename
         }
 
-        params.update(self._auth_params())
+        if multipart_upload:
+            params['multipart_upload'] = True
```
Comment (manneohrstrom) [inline shotgun_api3/shotgun.py]: double quotes pls

File: shotgun_api3/shotgun.py [hunk 7]
```diff
@@ -2342,12 +2321,13 @@ def _upload_to_sg(self, entity_type, entity_id, path, field_name, display_name,
...
-    def _get_attachment_upload_info(self, is_thumbnail, filename):
+    def _get_attachment_upload_info(self, is_thumbnail, filename, multipart_upload):
         """
         Internal function to get the information needed to upload a file to Cloud storage.
 
         :param bool is_thumbnail: indicates if the attachment is a thumbnail.
         :param str filename: name of the file that will be uploaded.
+        :param bool multipart_upload: Indicates if we want multi-part upload information back.
 
         :returns: dictionary containing the upload url and
```
Comment (manneohrstrom) [inline shotgun_api3/shotgun.py]: please update the returns docstring to reflect the new fields we are returning

File: shotgun_api3/shotgun.py [hunk 8]
```diff
@@ -2342,12 +2321,13 @@ def _upload_to_sg(self, entity_type, entity_id, path, field_name, display_name,
...
         return attachment_id
 
-    def _get_attachment_upload_info(self, is_thumbnail, filename):
+    def _get_attachment_upload_info(self, is_thumbnail, filename, multipart_upload):
         """
         Internal function to get the information needed to upload a file to Cloud storage.
 
         :param bool is_thumbnail: indicates if the attachment is a thumbnail.
         :param str filename: name of the file that will be uploaded.
+        :param bool multipart_upload: Indicates if we want multi-part upload information back.
```
Comment (manneohrstrom) [inline shotgun_api3/shotgun.py]: minor: might be worth explaining a bit more around what the difference is - does the return data change depending on this flag?

File: shotgun_api3/shotgun.py [hunk 9]
```diff
@@ -2374,11 +2355,17 @@ def _get_attachment_upload_info(self, is_thumbnail, filename):
         if not str(upload_info).startswith("1"):
             raise ShotgunError("Could not get upload_url but " \
                                "not sure why.\nPath: %s\nUrl: %s\nError: %s" % (
-                                   path, url, str(upload_info)))
+                                   filename, url, str(upload_info)))
 
         LOG.debug("Completed rpc call to %s" % (upload_url))
+
+        upload_info_parts = str(upload_info).split("\n")
```
Comment (manneohrstrom) [inline shotgun_api3/shotgun.py]: do we need to worry about different versions of the sg server responding in different ways here? or is it consistent - not sure if the server side changes are going out as a patch release or if they are perhaps already there. If they are going out in a patch, we might have to add checks to handle different server responses differently?

File: shotgun_api3/shotgun.py [hunk 10]
```diff
@@ -3539,6 +3526,137 @@ def _dict_to_extra_data(self, d, key_name="value"):
...
+        try:
+            content_type = mimetypes.guess_type(path)[0]
+            content_type = content_type or "application/octet-stream"
+            file_size = os.fstat(fd.fileno())[stat.ST_SIZE]
+            filename = os.path.basename(path)
+
+            etags = []
+            part_number = 1
+            bytes_read = 0
+            chunk_size = 20000000
```
Comment (manneohrstrom) [inline shotgun_api3/shotgun.py]: could we reuse that constant here?

File: shotgun_api3/shotgun.py [hunk 11]
```diff
@@ -3539,6 +3526,137 @@ def _dict_to_extra_data(self, d, key_name="value"):
...
+            fd.close()
+
+        LOG.debug("File uploaded to Cloud storage: %s", filename)
+
+    def _multipart_upload_file_to_storage(self, path, upload_info):
+        """
+        Internal function to upload a file to the Cloud storage in multiple parts.
+
+        :param str path: Full path to an existing non-empty file on disk to upload.
+        :param dict upload_info: Contains details received from the server, about the upload.
```
Comment (manneohrstrom) [inline shotgun_api3/shotgun.py]: would it helpful to document the expected contents of this dict, at least in one of the methods? It looks like we are accessing specific keys, so it's not just a case of sending it blindly back to the server. If we document it in one place there rest of the docstring could point to that one i guess....

File: shotgun_api3/shotgun.py [hunk 12]
```diff
@@ -3539,6 +3526,137 @@ def _dict_to_extra_data(self, d, key_name="value"):
...
+            bytes_read = 0
+            chunk_size = 20000000
+            while bytes_read < file_size:
+                data = fd.read(chunk_size)
+                bytes_read += len(data)
+                part_url = self._get_upload_part_link(upload_info, filename, part_number)
+                etags.append(self._upload_data_to_storage(data, content_type, len(data), part_url ))
+                part_number += 1
+
+            self._complete_multipart_upload(upload_info, filename, etags)
```
Comment (manneohrstrom) [inline shotgun_api3/shotgun.py]: what happens if it fails half way? Should we worry about that case? should we send a message to the server saying that things broke half way and to discard the upload? Or will the server simply discard any "dangling" pieces after a certain amount of idle time?

I think a piece of DX to consider (and point out in the code) here is that we want a user to get the same experience when a multi-part upload fails and when a normal upload fails. Is this currently the case? Would be great with some tests to assert this.

I am thinking about the specific case where clients have flaky connections (not uncommon) so one data exchange with the server may work fine but the next one fails due to external factors.

File: shotgun_api3/shotgun.py [hunk 13]
```diff
@@ -3539,6 +3526,137 @@ def _dict_to_extra_data(self, d, key_name="value"):
...
+
+        :param dict upload_info: Contains details received from the server, about the upload.
+        :param str filename: Name of the file for which we want the link.
+        :param int part_number: Part number for the link.
+        :returns: upload url.
+        :rtype: str
+        """
+        params = {
+            "upload_type": upload_info["upload_type"],
+            "filename": filename,
```
Comment (manneohrstrom) [inline shotgun_api3/shotgun.py]: curious - why is the filename important here? i am guessing that the server relies on the `upload_id` to piece things together... why do we need to send the filename as well?

might be worth adding some comments to explain

File: shotgun_api3/shotgun.py [hunk 14]
```diff
@@ -3539,6 +3526,137 @@ def _dict_to_extra_data(self, d, key_name="value"):
...
+        Internal function to upload data to Cloud storage.
+        
+        :param stream data: Contains details received from the server, about the upload.
+        :param str content_type: Content type of the data stream.
+        :param int size: Number of bytes in the data stream.
+        :param str storage_url: Target URL for the uploaded file. 
+        :returns: upload url.
+        :rtype: str
+        """
+        try:
```
Comment (manneohrstrom) [inline shotgun_api3/shotgun.py]: logging here would be great

File: shotgun_api3/shotgun.py [hunk 15]
```diff
@@ -3539,6 +3526,137 @@ def _dict_to_extra_data(self, d, key_name="value"):
...
+            "timestamp": upload_info["timestamp"],
+            "upload_id": upload_info["upload_id"],
+            "etags": ",".join(etags)
+        }
+
+        url = urlparse.urlunparse((self.config.scheme, self.config.server,
+                                   "/upload/api_complete_multipart_upload", None, None, None))
+        result = self._send_form(url, params)
+
+        if not str(result).startswith("1"):
```
Comment (manneohrstrom) [inline shotgun_api3/shotgun.py]: please add comments explaining what the response looks like

File: shotgun_api3/shotgun.py (hunk 4 above)
Comment (daigles) [inline shotgun_api3/shotgun.py]: done!

File: shotgun_api3/shotgun.py (hunk 8 above)
Comment (daigles) [inline shotgun_api3/shotgun.py]: I intentionally kept this somewhat opaque to help keep users away from the details...

File: shotgun_api3/shotgun.py (hunk 9 above)
Comment (daigles) [inline shotgun_api3/shotgun.py]: we're good! only 1 version of the server-side code is used!

File: shotgun_api3/shotgun.py (hunk 11 above)
Comment (daigles) [inline shotgun_api3/shotgun.py]: My intention is really to keep the details of the 'upload_info' as opaque as possible.
This info is used by private methods managing the upload and users should not be messing with these...

File: shotgun_api3/shotgun.py (hunk 13 above)
Comment (daigles) [inline shotgun_api3/shotgun.py]: server needs both the filename and upload id...

File: shotgun_api3/shotgun.py (hunk 9 above)
Comment (manneohrstrom) [inline shotgun_api3/shotgun.py]: sweet!

File: shotgun_api3/shotgun.py (hunk 11 above)
Comment (manneohrstrom) [inline shotgun_api3/shotgun.py]: makes sense!

**Pull request merged**